sdk/src/trace/batch_span_processor.cc

```diff
@@ -0,0 +1,285 @@
+      max_queue_size_(max_queue_size),
+      schedule_delay_millis_(schedule_delay_millis),
+      max_export_batch_size_(max_export_batch_size)
+{
+  buffer_ =
+      std::unique_ptr<CircularBuffer<Recordable>>(new CircularBuffer<Recordable>(max_queue_size_));
+
+  // Start the background worker thread
+  worker_thread_ =
+      std::unique_ptr<std::thread>(new std::thread(&BatchSpanProcessor::DoBackgroundWork, this));
+}
+
+std::unique_ptr<Recordable> BatchSpanProcessor::MakeRecordable() noexcept
+{
+  return exporter_->MakeRecordable();
+}
+
+void BatchSpanProcessor::OnStart(Recordable &span) noexcept
+{
+  // no-op
+  (void)span;
+}
+
+void BatchSpanProcessor::OnEnd(std::unique_ptr<Recordable> &&span) noexcept
+{
+  if (is_shutdown_ == true)
+  {
+    // TODO: log in glog
+    return;
+  }
+
+  std::unique_lock<std::mutex> lk(cv_m_);
+
+  if (buffer_->Add(span) == false)
+  {
+    // TODO: glog that spans will likely be dropped
+  }
+
+  // If the queue gets at least half full a preemptive notification is
+  // sent to the worker thread to start a new export cycle.
+  if (buffer_->size() >= max_queue_size_ / 2)
+  {
+    // signal the worker thread
+    cv_.notify_one();
+  }
+}
+
+void BatchSpanProcessor::ForceFlush(std::chrono::microseconds timeout) noexcept
+{
+  if (is_shutdown_ == true)
+  {
+    // TODO: glog that processor is already shutdown
+  }
+
+  is_force_flush_ = true;
+
+  std::chrono::microseconds timeout_left = timeout;
+
+  // Keep attempting to wake up the worker thread and monitor timeout left
+  // accordingly.
+  while (is_force_flush_ == true && timeout_left.count() > 0)
+  {
+    auto start = std::chrono::steady_clock::now();
+    cv_.notify_one();
+    auto end = std::chrono::steady_clock::now();
+    timeout_left -= std::chrono::duration_cast<std::chrono::microseconds>(end - start);
+  }
+
+  // Now wait for the worker thread to signal back from the Export method
+  //                      OR
+  // Wait until we timeout
+  std::unique_lock<std::mutex> lk(force_flush_cv_m_);
+  while (timeout_left.count() > 0 && is_force_flush_notified_ == false)
+  {
+    auto start_time = std::chrono::steady_clock::now();
+    force_flush_cv_.wait_for(lk, timeout_left);
+    timeout_left -= std::chrono::duration_cast<std::chrono::microseconds>(
+        std::chrono::steady_clock::now() - start_time);
+  }
+
+  lk.unlock();
+
+  // This flag handles the case when we have a timeout too because the notification
+  // loop (lines 202 - 205) in the Export method keeps looping until is_force_flush_notified_
+  // holds true. As soon as this flag is turned to false, the notification loop breaks.
+  is_force_flush_notified_ = false;
+
+  if (timeout_left.count() <= 0)
+  {
+    // TODO: glog timeout
+  }
+  else
+  {
+    // TODO: glog no timeout
+  }
+}
+
+void BatchSpanProcessor::DoBackgroundWork()
+{
+  auto timeout = schedule_delay_millis_;
+
+  while (true)
+  {
+    std::unique_lock<std::mutex> lk(cv_m_);
```

Comment:
We hold this lock for the whole duration of `wait_for` below?

Reply:
The lock is released when we start waiting.